x/gov/proposal.go

```diff
@@ -0,0 +1,180 @@
+func (keeper Keeper) DeleteProposal(ctx sdk.Context, proposalID uint64) {
+	store := ctx.KVStore(keeper.storeKey)
+	proposal, ok := keeper.GetProposal(ctx, proposalID)
+	if !ok {
+		panic("DeleteProposal cannot fail to GetProposal")
+	}
+	keeper.RemoveFromInactiveProposalQueue(ctx, proposalID, proposal.DepositEndTime)
+	keeper.RemoveFromActiveProposalQueue(ctx, proposalID, proposal.VotingEndTime)
+	store.Delete(ProposalKey(proposalID))
+}
+
+// GetProposals returns all the proposals from store
+func (keeper Keeper) GetProposals(ctx sdk.Context) (proposals Proposals) {
+	keeper.IterateProposals(ctx, func(proposal types.Proposal) bool {
+		proposals = append(proposals, proposal)
+		return false
+	})
+	return
+}
+
+// GetProposalsFiltered get Proposals from store by ProposalID
+// voterAddr will filter proposals by whether or not that address has voted on them
+// depositorAddr will filter proposals by whether or not that address has deposited to them
+// status will filter proposals by status
+// numLatest will fetch a specified number of the most recent proposals, or 0 for all proposals
+func (keeper Keeper) GetProposalsFiltered(ctx sdk.Context, voterAddr sdk.AccAddress, depositorAddr sdk.AccAddress, status ProposalStatus, numLatest uint64) []Proposal {
+
+	maxProposalID, err := keeper.peekCurrentProposalID(ctx)
+	if err != nil {
+		return nil
+	}
+
+	matchingProposals := []Proposal{}
+
+	if numLatest == 0 {
+		numLatest = maxProposalID
+	}
+
+	for proposalID := maxProposalID - numLatest; proposalID < maxProposalID; proposalID++ {
+		if voterAddr != nil && len(voterAddr) != 0 {
+			_, found := keeper.GetVote(ctx, proposalID, voterAddr)
+			if !found {
+				continue
+			}
+		}
+
+		if depositorAddr != nil && len(depositorAddr) != 0 {
+			_, found := keeper.GetDeposit(ctx, proposalID, depositorAddr)
+			if !found {
+				continue
+			}
+		}
+
+		proposal, ok := keeper.GetProposal(ctx, proposalID)
+		if !ok {
+			continue
+		}
+
+		if ValidProposalStatus(status) && proposal.Status != status {
+			continue
+		}
+
+		matchingProposals = append(matchingProposals, proposal)
+	}
+	return matchingProposals
+}
+
+// GetLastProposalID gets the last used proposal ID. Returns 0 if the current proposal ID is 0
+// or if the initial proposal ID hasn't been set.
+func (keeper Keeper) GetLastProposalID(ctx sdk.Context) (proposalID uint64) {
+	proposalID, err := keeper.peekCurrentProposalID(ctx)
+	if err != nil || proposalID == 0 {
+		return 0
+	}
+	return proposalID - 1
+}
+
+// Gets the next available ProposalID and increments it
+func (keeper Keeper) getNewProposalID(ctx sdk.Context) (proposalID uint64, err sdk.Error) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get(NextProposalIDKey)
+	if bz == nil {
+		return 0, ErrInvalidGenesis(keeper.codespace, "initial proposal ID never set")
+	}
+	keeper.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &proposalID)
+	bz = keeper.cdc.MustMarshalBinaryLengthPrefixed(proposalID + 1)
+	store.Set(NextProposalIDKey, bz)
```

Comment:
I'm not 100% sure it's safe to set it in the store inside a getter, I'll try to elaborate in SubmitProposal

Reply:
So basically I think it's safer to have the setter separate from the getter